go/protomodule/protomodule_message.go

```diff
@@ -0,0 +1,370 @@
+	// use without initialization works.
+	//
+	// Example:
+	//   msg = MyProtoMessage()
+	//   msg.repeated_field.append("a")
+	//   # msg.repeated_field should be ["a"]
+	if fieldDesc.IsList() || fieldDesc.IsMap() || fieldDesc.Kind() == protoreflect.MessageKind {
+		msg.SetField(name, starlarkValue)
+	}
+
+	return starlarkValue, nil
+}
+
+func (msg *protoMessage) AttrNames() []string {
+	return fieldNames(msg.msgDesc)
+}
+
+func fieldNames(msgDesc protoreflect.MessageDescriptor) []string {
+	out := []string{}
+	fields := msgDesc.Fields()
+	for i := 0; i < fields.Len(); i++ {
+		fieldName := string(fields.Get(i).Name())
+		out = append(out, fieldName)
+	}
+	sort.Strings(out)
+	return out
+}
+
+func (msg *protoMessage) SetField(name string, val starlark.Value) error {
+	fieldDesc := getFieldDescriptor(msg.msgDesc, name)
+	if fieldDesc == nil {
+		return fmt.Errorf("AttributeError: `%s' value has no field %q", msg.Type(), name)
+	}
+
+	if err := msg.CheckMutable("set field of"); err != nil {
+		return err
+	}
+
+	// Convert starlark.List and starlark.Dict on assignment
+	if fieldDesc.IsList() {
+		if starlarkListVal, ok := val.(*starlark.List); ok {
+			// Convert starlark.List to protoRepeated
+			list, err := newProtoRepeatedFromList(fieldDesc, starlarkListVal)
+			if err != nil {
+				return err
+			}
+
+			val = list
+		}
+	} else if fieldDesc.IsMap() {
+		if starlarkDictVal, ok := val.(*starlark.Dict); ok {
+			// Convert stalark.Map into protoMap
+			mapVal, err := newProtoMapFromDict(fieldDesc.MapKey(), fieldDesc.MapValue(), starlarkDictVal)
+			if err != nil {
+				return err
+			}
+
+			val = mapVal
+		}
+	}
+
+	// Allow using msg_field = None to unset a scalar message field
+	if fieldDesc.Kind() == protoreflect.MessageKind && val == starlark.None && !fieldDesc.IsList() && !fieldDesc.IsMap() {
+		delete(msg.fields, name)
+		return nil
+	}
+
+	// If valueFromStarlark returns an error, the val cannot be assigned to the field
+	_, err := valueFromStarlark(msg.msg.ProtoReflect(), fieldDesc, val)
+	if err != nil {
+		return err
+	}
+
+	// Clear other oneof
+	oneof := fieldDesc.ContainingOneof()
+	if oneof != nil {
+		fields := oneof.Fields()
+		for i := 0; i < fields.Len(); i++ {
+			delete(msg.fields, string(fields.Get(i).Name()))
+		}
+	}
+
+	msg.fields[name] = val
+
+	return nil
+}
+
+// Applies default values for proto2 fields with defaults that are not already set
+func (msg *protoMessage) SetDefaults() error {
+	if err := msg.CheckMutable("set field defaults of"); err != nil {
+		return err
+	}
+
+	for _, fieldName := range fieldNames(msg.msgDesc) {
+		fieldDesc := getFieldDescriptor(msg.msgDesc, fieldName)
+		if fieldDesc == nil {
+			continue
+		}
+
+		// Already set, do not set default
+		if _, ok := msg.fields[fieldName]; ok {
+			continue
+		}
+
+		if !fieldDesc.HasDefault() {
+			continue
+		}
+
+		val, err := valueToStarlark(fieldDesc.Default(), fieldDesc)
+		if err != nil {
```

Comment:
OOC - would we want to revert the fields for which we've already set defaults? Or is it okay to leave the proto message in a partially updated state where some fields have defaults and others don't (but could)?

Relatedly, is there value in collecting errors (to return at the end of the method) while continuing to set as many default values as possible?

Reply:
Are you asking if `SetDefaults` should behave like first clearing the message like `proto.Reset` and then setting the defaults for fields that have them? If so this is what I would expect and that's what the stated behavior in the docs seems to apply https://github.com/stripe/skycfg/blob/trunk/docs/modules.asciidoc#proto.set_defaults but let me look into how the current implementation behaves

With errors, I don't think it really matters as when an error is returned the starlark/skycfg evaluation stops and returns the error. Since Skycfg is not mutating any values outside of the evaluation, all these values will just get thrown away anyway even if in a partial state

Reply:
Alright I tested the existing implementation and it does not behave this way, only setting values with defaults not clearing those without. It calls out to https://github.com/golang/protobuf/blob/master/proto/defaults.go#L25 and here it's filtering out to only fields with `HasDefault()` which AFAICT is only true for fields with proto2 defaults

Reply:
Sorry for the confusion!

> Are you asking if SetDefaults should behave like first clearing the message like proto.Reset and then setting the defaults for fields that have them?

No, I was asking "when `SetDefaults` errors, should we clear the errors that were successfully set earlier in `SetDefaults`?" i.e. Should setting defaults be all-or-nothing or "best effort"?

But based on
> Since Skycfg is not mutating any values outside of the evaluation, all these values will just get thrown away anyway even if in a partial state
it sounds like this is all-or-nothing?

>That said I would prefer to match the existing behavior as closely as possible to de-risk the new implementation refactor, and then change any behavior separately/explicitly afterwards
👍 SGTM!